File: ICPC/2015 Divisionals/atomic/submissions/atomic_darcy.py
import re

n = int(input())

elems = dict()

for _ in range(n):
  elem, mass = input().split()
  elems[elem] = int(mass.replace(".",""))

compound = re.findall('[A-Z][a-z0-9]*',input())

ans = 0
for part in compound:
  x = re.findall('\d+|\D+', part)
  elem = x[0]
  num = 1 if len(x) == 1 else int(x[1])
  ans += elems[elem] * num

print(ans / 100.0)

File: ICPC/2015 Divisionals/atomic/submissions/AtomicDarcy.java
import java.util.Scanner;

public class AtomicDarcy {
  public static void main(String[] args) {
    new AtomicDarcy().run();
  }

  void run() {
    Scanner in = new Scanner(System.in);
    int n = in.nextInt();
    String[] elems = new String[n];
    int[] mass = new int[n];

    for(int i=0; i<n; i++) {
      elems[i] = in.next();
      String myMass = in.next();
      for(int j=0; j<myMass.length(); j++)
        if(myMass.charAt(j) != '.')
          mass[i] = 10*mass[i] + (myMass.charAt(j)-'0');

    }
    String compound = in.next();

    int ans = 0;
    for(int i=0; i<compound.length(); i++) {
      String elem = compound.substring(i,i+1);
      if(i+1 < compound.length() && Character.isLowerCase(compound.charAt(i+1))) {
        elem += compound.charAt(i+1);
        i++;
      }
      int num = 0;
      while(i+1 < compound.length() && Character.isDigit(compound.charAt(i+1))) {
        num = 10*num + (compound.charAt(i+1)-'0');
        i++;
      }
      if(num == 0) num = 1;
      for(int j=0; j<n; j++)
        if(elem.equals(elems[j]))
          ans += mass[j] * num;
    }
    System.out.println(ans / 100.0);
  }
}
